In [2]:
# Importing the neccessary libraries
import pandas as pd
import numpy as np
import statistics as stats
import os
import keras
from keras.models import Sequential
from keras.layers import Dense
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from datetime import datetime

# Forcing keras to use CPU.
os.environ["CUDA_VISIBLE_DEVICES"] = "-1"

In [6]:
# Reading the Data and storing it in a dataframe

df = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
df.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Fine Aggregate  Age  Strength
0   540.0                 0.0      0.0  ...           676.0   28     79.99
1   540.0                 0.0      0.0  ...           676.0   28     61.89
2   332.5               142.5      0.0  ...           594.0  270     40.27
3   332.5               142.5      0.0  ...           594.0  365     41.05
4   198.6               132.4      0.0  ...           825.5  360     44.30

[5 rows x 9 columns]

In [9]:
print('\nShape of dataframe : ',df.shape)


Shape of dataframe :  (1030, 9)


In [10]:
# Summary of the dataset
df.describe()

Cement  Blast Furnace Slag  ...          Age     Strength
count  1030.000000         1030.000000  ...  1030.000000  1030.000000
mean    281.167864           73.895825  ...    45.662136    35.817961
std     104.506364           86.279342  ...    63.169912    16.705742
min     102.000000            0.000000  ...     1.000000     2.330000
25%     192.375000            0.000000  ...     7.000000    23.710000
50%     272.900000           22.000000  ...    28.000000    34.445000
75%     350.000000          142.950000  ...    56.000000    46.135000
max     540.000000          359.400000  ...   365.000000    82.600000

[8 rows x 9 columns]

<b>Note </b> : As the data is the same as used in this course, the video lectures have already shown that there is no need to perform any kind of pre-processing on the data. Hence, even the most common checks of data such as checking for missing values are not performed.

In [11]:
# Creating dataframes of features (X) and target (Y)
X = df.iloc[:, 0:8]
Y = df.iloc[:,8]

# Printing the dataframes X and Y to ensure we have created the dataframes with the correct columns
print('The features or the predictors (X) are : ', X, '\n\n') 
print('The target (Y) is : ', Y, '\n\n')

The features or the predictors (X) are :        Cement  Blast Furnace Slag  ...  Fine Aggregate  Age
0      540.0                 0.0  ...           676.0   28
1      540.0                 0.0  ...           676.0   28
2      332.5               142.5  ...           594.0  270
3      332.5               142.5  ...           594.0  365
4      198.6               132.4  ...           825.5  360
...      ...                 ...  ...             ...  ...
1025   276.4               116.0  ...           768.3   28
1026   322.2                 0.0  ...           813.4   28
1027   148.5               139.4  ...           780.0   28
1028   159.1               186.7  ...           788.9   28
1029   260.9               100.5  ...           761.5   28

[1030 rows x 8 columns] 


The target (Y) is :  0       79.99
1       61.89
2       40.27
3       41.05
4       44.30
        ...  
1025    44.28
1026    31.18
1027    23.70
1028    32.77
1029    32.40
Name: Strength, Length: 1030, dtype: float64 



<b>Note 1</b> : Unlike the method in this course, the splitting is done using indexing instead of using the names of the columns. Additionally, a different notation is used. The word <i>features</i> is used instead of <i>predictors</i>.


<b>Note 2</b> : Pandas indexes columns starting from 0. Note in the code below for the features (X) indexing is used as `[:, 0:8]`. The first part preceding the coma `(:)` tells pandas to include ALL rows of the original dataframe (df) in the new dataframe called X while the part succedding the comma `(0:8)` tells pandas to include all columns of the original dataframe (df) starting from column with index = 0 and ending with column with index = 7, <b> but not to include the column with index = 8 </b>  

<b>Note 3</b> : In order to split the data into train and test sets, the train_test_split function of the sklearn library is used. `The random_state` is used to ensure that the train and test split is the same each time, i.e. the train set and the test set have the same samples each time the code is run which is good for reproducing the results. If left empty, the random state is used by `np.random`. Since the Project requires splitting data the into <b>random</b> sets, hence `random_state` is not used, i.e. no value is set for random state. As the data hase to be split randomly  into train and test sets <b>50</b> times, a for loop will be used to to split the data in train test sets for <b>each model</b> 

In [12]:
def regression_model() :
    
    # Create the model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(X.shape[1],)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))

    # Compile the model
    model.compile(optimizer='adam', loss='mean_squared_error')
    
    return model

In [14]:
def data_split() :
    X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3)
    
    # Create a list containing X_train, X_test, Y_train, Y_test and return the list
    splits = [X_train, X_test, Y_train, Y_test] 
    return splits

<b>Note </b> : In the above function `data_split()` the <i>X_train, X_test, Y_train, Y_test</i> sets are stored in a list and list is returned. This is to ensure that the <i>X_train, X_test, Y_train, Y_test</i> sets are not printed when the function is called

In [15]:
def predict() :
    return model.predict(X_test)

In [16]:
def calculate_mse() :
    return mean_squared_error(Y_test,Y_predicted)

<b>Note 1</b> : The function `regression_model` as defined above only **compiles** the model and doesn't fit the model to training set. This is because in PART C, the number of epochs are changed and this would allow to fit the model with new number of epochs. This function will be used for **PART A**, **PART B** and **PART C**.

<b>Note 2</b> : Since the splitting data, prediction and mean squared error calculations are all being used repeatly as well, hence there are separate functions created for them to remove redundancy of typing the same line of code while also making the code more neat. However,  as we the features (X) are to be normalized only **once**, hence there is no need to create a function for it

<b>Note 3</b> : As the split function is executed first, the training and test sets are obtained and hence there is no need to explicitly pass any arguments to `predict()` and `calculate_mse()` functions. 

# <font color = blue> PART A : BUILDING A BASELINE MODEL </font>


<b>The baseline model consists of the following : </b>
    <ul>
        <li> Input layer with 10 nodes </li>
        <li> A single hidden layer with 10 nodes and ReLU activation function </li>
        <li> Adam optimizer and mean squared error loss function </li>
    </ul>

<p/>

## <font color = darkorange>Task 1 : Train and Test the Baseline Model</font>

In order to train and test the the baseline model, the following steps are performed :
<ol>
    <li>Randomly split the data into <i>X_train, X_test, Y_train, Y_test</i> sets</li>
    <li>Train the model using <i>X_train, Y_train</i> using <b>50</b> epochs</li>
    <li>Get the predictions on the <i>X_test</i> set </li>
    <li>Compute the <b>mean squared error</b> on the test set using sklearn</li>
</ol>

### <font color = #2980B9> Step 1 : Randomly split the data into <i>X_train, X_test, Y_train, Y_test </font>

In [17]:
# Split data into X_train, X_test, Y_train, Y_test
X_train, X_test, Y_train, Y_test = data_split()

### <font color = #2980B9> Step 2 : Train the model using <i>X_train, Y_train</i> using 50 epochs </font>

In [18]:
# Create the model
model = regression_model()

# Fit the model on the train set
model.fit(X_train, Y_train, validation_split=0.3, epochs=50)

Epoch 1/50
16/16 [==============================] - 1s 32ms/step - loss: 607.5514 - val_loss: 418.4524
Epoch 2/50
16/16 [==============================] - 0s 4ms/step - loss: 397.4071 - val_loss: 347.5311
Epoch 3/50
16/16 [==============================] - 0s 3ms/step - loss: 332.0757 - val_loss: 293.2103
Epoch 4/50
16/16 [==============================] - 0s 4ms/step - loss: 270.6738 - val_loss: 250.0178
Epoch 5/50
16/16 [==============================] - 0s 4ms/step - loss: 246.1282 - val_loss: 219.0081
Epoch 6/50
16/16 [==============================] - 0s 4ms/step - loss: 198.5568 - val_loss: 196.0515
Epoch 7/50
16/16 [==============================] - 0s 4ms/step - loss: 184.3123 - val_loss: 178.0841
Epoch 8/50
16/16 [==============================] - 0s 4ms/step - loss: 158.0628 - val_loss: 167.1625
Epoch 9/50
16/16 [==============================] - 0s 4ms/step - loss: 138.6251 - val_loss: 157.8757
Epoch 10/50
16/16 [==============================] - 0s 4ms/step - loss: 148.8151

### <font color = #2980B9> Step 3 : Get the predictions on the X_test </font>

In [19]:
# Store the predictions in a variable Y_Predicted
Y_predicted = predict()

<b>Note </b> : Y_test or the original values are also sometimes refered to as Y_true and this is the notation used in the examples found on the [mean square error page of sklearn](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_squared_error.html). However, in this notebook the Y_test notations is used for original values 

### <font color = #2980B9> Step 4 : Compute the <i>mean squared error</i> on the test set using sklearn </font>

In [20]:
# Calculate the mean square error
mse = calculate_mse()
#mse = mean_squared_error(Y_test,Y_predicted)
print('Mean Square Error (MSE) of the Baseline Model is : ' , mse)

Mean Square Error (MSE) of the Baseline Model is :  88.39285073988803


## <font color = darkorange>Task 2 : Create 50 Models and Calculate the <i>µ</i> & <i>σ</i> of their MSE Errors</font>

In order to train 50 models and calulate the mean (µ) and standard deviation (σ) of their mean square errors (MSE)  :
<ol>
    <li>Create an empty list <code>list_of_mse</code> to store the mean square error of each of the models</li>
    <li>Define a for loop and perform each of the following steps in the loop</li>
        <ol>
        <li>Randomly split the data into <i>X_train, X_test, Y_train, Y_test</i> sets</li>
        <li>Train the model using <i>X_train, Y_train</i> using <b>50</b> epochs</li>
        <li>Evaluate the model using <i>X_test, Y_test</i></li>
        <li>Get the predictions on the <i>X_test</i> set </li>
        <li>Compute the <b>mean squared error</b> on the test set using sklearn</li>
    </ol>
</ol>

<b>Note</b> : To calcuate the mean and standard deviation of the mean square errors (mse) of 50 models which are stored in <code>list_of_means</code>, I will be using the python library <code>statistics</code> which has builtin functions to help caluclate the mean and standard deviation of a list

### <font color = #2980B9> Step 1 : Creating the <i>list_of_mse</i> list</font>

In [21]:
# Create the empty lists
list_of_mse = []

### <font color = #2980B9> Step 2 : Creating 50 Models, Calculating The MSE for Each and <i>µ</i> & <i>σ</i> of the 50 MSE Values in <i>list_of_mse</i> </font>

In [22]:
# Create the for loop to split the data, create, compile & fit model, evaluate & nake predictions, caluclate mse and store
# in list_of_mse

start_time = datetime.now() # Starting time of the for loop execution

for i in range(50) :
    # Split the data into train and test set
    data_split()
    
    # Create and compile the regression model using the function regression_model as defined in TASK 1
    model = regression_model()

    # Fit the model on the train set
    print('\n\n\nTraining Model # ' , i+1 , '\n\n') # Print the Model Number that is being trained
    model.fit(X_train, Y_train, validation_split=0.3, epochs=50)
    print('\n')
    
    # Make prediction on the test set
    Y_predicted = predict()
    
    # Calculate the mean square error
    mse = calculate_mse()
    print('Mean Squared Error for Training Model # ', i+1 , ' : ', mse)
    
    # Add the mse to the list_of_mse list
    list_of_mse.append(mse)

end_time = datetime.now() # Ending time of the for loop execution

# Print time taken for fitting 50 models and calucating the Mean and Standard Deviation of MSE of 50 models
print('\n\nTotal Execution Time : ' , format(end_time - start_time))
    

Streaming output truncated to the last 5000 lines.
16/16 [==============================] - 0s 4ms/step - loss: 632.1438 - val_loss: 508.2395
Epoch 7/50
16/16 [==============================] - 0s 4ms/step - loss: 471.7532 - val_loss: 456.2940
Epoch 8/50
16/16 [==============================] - 0s 3ms/step - loss: 469.7366 - val_loss: 420.8598
Epoch 9/50
16/16 [==============================] - 0s 4ms/step - loss: 461.0926 - val_loss: 395.2240
Epoch 10/50
16/16 [==============================] - 0s 5ms/step - loss: 404.3616 - val_loss: 364.9857
Epoch 11/50
16/16 [==============================] - 0s 4ms/step - loss: 348.8219 - val_loss: 344.0170
Epoch 12/50
16/16 [==============================] - 0s 4ms/step - loss: 318.0613 - val_loss: 325.5320
Epoch 13/50
16/16 [==============================] - 0s 4ms/step - loss: 327.4472 - val_loss: 307.3285
Epoch 14/50
16/16 [==============================] - 0s 4ms/step - loss: 323.8883 - val_loss: 290.9370
Epoch 15/50
16/16 [==================

In [23]:
# Calculate the Mean of the MSE of 50 models
mean_of_mse = stats.mean(list_of_mse)

# Calculate the Standard Deviation of the MSE of 50 models
std_of_mse = stats.stdev(list_of_mse)

# Print the Mean and Standard Deviation of MSE of 50 models
print('\n\nMean of the MSE of 50 Models : ' , str(mean_of_mse))
print('Standard Deviation of MSE of 50 Models : ' , str(std_of_mse))



Mean of the MSE of 50 Models :  212.12918204304935
Standard Deviation of MSE of 50 Models :  136.03460413795105


<br/>

# <font color = blue> END OF PART A </font>

<hr/>

# <font color = blue> PART B : BASELINE MODEL WITH NORMALIZED FEATURES </font>


In this part, all the tasks from <b>PART A</b> are performed, but this time the values for the features (X) will be normalized using the formula `X - µ / σ`



## <font color = darkorange>Task 1 : Train and Test the Baseline Model with Normalized Features</font>

In order to train and test the the baseline model with normalized features, the following steps are performed :
<ol>
    <li>Normalize the features (X)</li>
    <li>Randomly split the data into <i>X_train, X_test, Y_train, Y_test</i> sets</li>
    <li>Train the model using <i>X_train, Y_train</i> using <b>50</b> epochs</li>
    <li>Get the predictions on the <i>X_test</i> set </li>
    <li>Compute the <b>mean squared error</b> on the test set using sklearn</li>
</ol>

### <font color = #2980B9> Step 1 : Normalize the features (X) </font>

In [24]:
X = X - X.mean() / X.std()

### <font color = #2980B9> Step 2 : Randomly split the data into <i>X_train, X_test, Y_train, Y_test </font>

In [25]:
# Creating X_train, X_test, Y_train and Y_test sets
X_train, X_test, Y_train, Y_test = data_split()

### <font color = #2980B9> Step 3 : Train the model using <i>X_train, Y_train</i> using 50 epochs </font>

<b>Note </b> : Since the regression model has already been created using `def regression_model` in <b>TASK 1</b> and there are no changes being made to it, hence there is no need to write the code for the model again. 

In [26]:
model = regression_model()

# Fit the model on the train set
model.fit(X_train, Y_train, validation_split=0.3, epochs=50)

Epoch 1/50
16/16 [==============================] - 1s 16ms/step - loss: 5194.8498 - val_loss: 2465.0786
Epoch 2/50
16/16 [==============================] - 0s 5ms/step - loss: 2623.0342 - val_loss: 1860.1479
Epoch 3/50
16/16 [==============================] - 0s 5ms/step - loss: 1919.7544 - val_loss: 1509.5568
Epoch 4/50
16/16 [==============================] - 0s 5ms/step - loss: 1582.6834 - val_loss: 1219.0430
Epoch 5/50
16/16 [==============================] - 0s 5ms/step - loss: 1207.4204 - val_loss: 975.2711
Epoch 6/50
16/16 [==============================] - 0s 4ms/step - loss: 957.2397 - val_loss: 786.6439
Epoch 7/50
16/16 [==============================] - 0s 5ms/step - loss: 805.4368 - val_loss: 632.8056
Epoch 8/50
16/16 [==============================] - 0s 5ms/step - loss: 668.1673 - val_loss: 479.7115
Epoch 9/50
16/16 [==============================] - 0s 5ms/step - loss: 437.1839 - val_loss: 345.4590
Epoch 10/50
16/16 [==============================] - 0s 5ms/step - loss:

### <font color = #2980B9> Step 4 : Get the predictions on the X_test </font>

In [27]:
# Store the predictions in a variable Y_Predicted
Y_predicted = predict()

### <font color = #2980B9> Step 5 : Compute the <i>mean squared error</i> on the test set using sklearn </font>

In [28]:
# Calculate the mean square error

mse = calculate_mse()
print('Mean Square Error (MSE) of the Baseline Model with Normalized Features is : ' , str(mse))

Mean Square Error (MSE) of the Baseline Model with Normalized Features is :  114.34176469410498


<br/>

<p/>

## <font color = darkorange>Task 2 : Create 50 Models and Calculate the <i>µ</i> & <i>σ</i> of their MSE Errors with new Features</font>

In order to train 50 models with new features (normalized features) and calulate the mean (µ) and standard deviation (σ) of their mean square errors (MSE) the following steps are performed :
<ol>
    <li>Create an empty list <code>list_of_mse</code> to store the mean square error of each of the models</li>
    <li>Define a for loop and perform each of the following steps in the loop</li>
        <ol>
        <li>Randomly split the data into <i>X_train, X_test, Y_train, Y_test</i> sets</li>
        <li>Train the model using <i>X_train, Y_train</i> using <b>50</b> epochs</li>
        <li>Evaluate the model using <i>X_test, Y_test</i></li>
        <li>Get the predictions on the <i>X_test</i> set </li>
        <li>Compute the <b>mean squared error</b> on the test set using sklearn</li>
    </ol>
</ol>

### <font color = #2980B9> Step 1 : Creating the <i>list_of_mse</i> list</font>

In [29]:
# Create the empty lists
list_of_mse = []

### <font color = #2980B9> Step 2 : Creating 50 Models, Calculating The MSE for Each and <i>µ</i> & <i>σ</i> of the 50 MSE Values in <i>list_of_mse</i> </font>

In [30]:
# Create the for loop to split the data, create, compile & fit model, evaluate & nake predictions, caluclate mse and store
# in list_of_mse

start_time = datetime.now() # Starting time of the for loop execution

for i in range(50) :
    # Split the data into train and test set
    X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3)
    
    # Create and compile the regression model using the function regression_model as defined in TASK 1
    model = regression_model()

    # Fit the model on the train set
    print('\n\n\nTraining Model # ' , i+1 , '\n\n') # Print the Model Number that is being trained
    model.fit(X_train, Y_train, validation_split=0.3, epochs=50)
    print('\n')
    
    # Make prediction on the test set
    Y_predicted = model.predict(X_test)
    
    # Calculate the mean square error
    mse = mean_squared_error(Y_test, Y_predicted)
    
    # Add the mse to the list_of_mse list
    list_of_mse.append(mse)

end_time = datetime.now() # Ending time of the for loop execution

# Print time taken for fitting 50 models and calucating the Mean and Standard Deviation of MSE of 50 models
print('\n\nTotal Execution Time : ' , format(end_time - start_time))
    

Streaming output truncated to the last 5000 lines.
16/16 [==============================] - 0s 4ms/step - loss: 175.9152 - val_loss: 197.8741
Epoch 38/50
16/16 [==============================] - 0s 5ms/step - loss: 186.1715 - val_loss: 192.7168
Epoch 39/50
16/16 [==============================] - 0s 4ms/step - loss: 169.3052 - val_loss: 190.5627
Epoch 40/50
16/16 [==============================] - 0s 5ms/step - loss: 160.5647 - val_loss: 185.8490
Epoch 41/50
16/16 [==============================] - 0s 4ms/step - loss: 181.3163 - val_loss: 182.0447
Epoch 42/50
16/16 [==============================] - 0s 4ms/step - loss: 161.1079 - val_loss: 181.1352
Epoch 43/50
16/16 [==============================] - 0s 4ms/step - loss: 158.3681 - val_loss: 173.1487
Epoch 44/50
16/16 [==============================] - 0s 4ms/step - loss: 151.8530 - val_loss: 180.7898
Epoch 45/50
16/16 [==============================] - 0s 4ms/step - loss: 147.3584 - val_loss: 166.8634
Epoch 46/50
16/16 [===============

In [31]:
# Calculate the Mean of the MSE of 50 models
mean_of_mse = stats.mean(list_of_mse)

# Calculate the Standard Deviation of the MSE of 50 models
std_of_mse = stats.stdev(list_of_mse)

# Print the Mean and Standard Deviation of MSE of 50 models
print('\n\nMean of the MSE of 50 Models : ' , str(mean_of_mse))
print('Standard Deviation of MSE of 50 Models : ' + str(std_of_mse))



Mean of the MSE of 50 Models :  184.57305818783047
Standard Deviation of MSE of 50 Models : 106.433563706641


<br/>

#### <font color = green> Comparision of Mean of MSE with Mean of MSE from PART A </font>
<p/>

<table style="width:20%">
  <tr>
    <th>Mean of MSE of PART A</th>
    <th>Mean of MSE of PART B</th>
  </tr>
  <tr>
    <td>212.13</td>
    <td>184.57</td>
  </tr>
</table>

The table above compares the Mean of **MSE for PART A** and **Mean of MSE for PART B**. As can be seen, the value of Mean of MSE of PART B is significantly smaller than that of PART A. The mean squared error tells how close a regression line is to a set of points. The smaller the value, the closer the model is to finding the line of best fit. So the smaller value for PART B shows that the model is slightly closer to finding a line of best fit.

To summarize, normalizing the features had a significant effect in reducing the MSE and finding the line of best fit

<b>Note</b> : Depending on the data, it may be impossible to get a very small value for MSE.

<br/>

# <font color = blue> END OF PART B </font>

<hr/>

<br/>

<br/>

<br/>

<hr/>

# <font color = blue> PART C : BASELINE MODEL WITH 100 EPOCHS </font>


In this part, all the tasks from <b>PART B</b> are performed, but this time the number of epochs are increased to 100



<p/>

<p/>

## <font color = darkorange>Task 1 : Train and Test the Baseline Model with Normalized Features and 100 epochs</font>

In order to train and test the the baseline model with normalized features and 100 epochs, the following steps are performed :
<ol>
    <li>Normalize the features (X)</li>
    <li>Randomly split the data into <i>X_train, X_test, Y_train, Y_test</i> sets</li>
    <li>Create a new model with 100 epochs</li>
    <li>Train the model using <i>X_train, Y_train</i> using <b>50</b> epochs</li>
    <li>Evaluate the model using <i>X_test, Y_test</i></li>
    <li>Get the predictions on the <i>X_test</i> set </li>
    <li>Compute the <b>mean squared error</b> on the test set using sklearn</li>
</ol>

### <font color = #2980B9> Step 1 : Normalize the features (X) </font>

<b>Note</b> : As the features (X) have already been normalized the features (X), hence this part is skipped

### <font color = #2980B9> Step 2 : Randomly split the data into <i>X_train, X_test, Y_train, Y_test </font>

In [32]:
# Creating X_train, X_test, Y_train and Y_test sets
X_train, X_test, Y_train, Y_test = data_split()

### <font color = #2980B9> Step 3 : Train the model using <i>X_train, Y_train</i> using 50 epochs </font>

<b>Note </b> : Since the regression model has already been created using `def regression_model` in <b>TASK 1</b> and there are no changes being made to it, hence there is no need to write the code for the model again. 

In [33]:
model = regression_model()

# Fit the model on the train set
model.fit(X_train, Y_train, validation_split=0.3, epochs=100)

Epoch 1/100
16/16 [==============================] - 1s 13ms/step - loss: 2609.7606 - val_loss: 1461.6296
Epoch 2/100
16/16 [==============================] - 0s 4ms/step - loss: 1281.3971 - val_loss: 1045.5585
Epoch 3/100
16/16 [==============================] - 0s 4ms/step - loss: 1038.1613 - val_loss: 867.0037
Epoch 4/100
16/16 [==============================] - 0s 4ms/step - loss: 795.5664 - val_loss: 623.7629
Epoch 5/100
16/16 [==============================] - 0s 4ms/step - loss: 581.6441 - val_loss: 527.3154
Epoch 6/100
16/16 [==============================] - 0s 4ms/step - loss: 433.4206 - val_loss: 436.7346
Epoch 7/100
16/16 [==============================] - 0s 3ms/step - loss: 400.3742 - val_loss: 370.1161
Epoch 8/100
16/16 [==============================] - 0s 3ms/step - loss: 346.2498 - val_loss: 323.5231
Epoch 9/100
16/16 [==============================] - 0s 3ms/step - loss: 269.7308 - val_loss: 296.3152
Epoch 10/100
16/16 [==============================] - 0s 3ms/step -

### <font color = #2980B9> Step 4 : Get the predictions on the X_test </font>

In [34]:
# Store the predictions in a variable Y_Predicted
Y_predicted = predict()

### <font color = #2980B9> Step 5 : Compute the <i>mean squared error</i> on the test set using sklearn </font>

In [35]:
# Calculate the mean square error

mse = calculate_mse()
print('Mean Square Error (MSE) of the Baseline Model with Normalized Features is : ' , str(mse))

Mean Square Error (MSE) of the Baseline Model with Normalized Features is :  114.68615748837777


<br/>

<p/>

## <font color = darkorange>Task 2 : Create 50 Models and Calculate the <i>µ</i> & <i>σ</i> of their MSE Errors with 100 Epochs</font>

In order to train 50 models with new features (normalized features) and calulate the mean (µ) and standard deviation (σ) of their mean square errors (MSE) the following steps are performed :
<ol>
    <li>Create an empty list <code>list_of_mse</code> to store the mean square error of each of the models</li>
    <li>Define a for loop and perform each of the following steps in the loop</li>
        <ol>
        <li>Randomly split the data into <i>X_train, X_test, Y_train, Y_test</i> sets</li>
        <li>Train the model using <i>X_train, Y_train</i> using <b>50</b> epochs</li>
        <li>Evaluate the model using <i>X_test, Y_test</i></li>
        <li>Get the predictions on the <i>X_test</i> set </li>
        <li>Compute the <b>mean squared error</b> on the test set using sklearn</li>
    </ol>
</ol>

### <font color = #2980B9> Step 1 : Creating the <i>list_of_mse</i> list</font>

In [36]:
# Create the empty lists
list_of_mse = []

### <font color = #2980B9> Step 2 : Creating 50 Models, Calculating The MSE for Each and <i>µ</i> & <i>σ</i> of the 50 MSE Values in <i>list_of_mse</i> </font>

In [37]:
# Create the for loop to split the data, create, compile & fit model, evaluate & nake predictions, caluclate mse and store
# in list_of_mse

start_time = datetime.now() # Starting time of the for loop execution

for i in range(50) :
    # Split the data into train and test set
    X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3)
    
    # Create and compile the regression model using the function regression_model as defined in TASK 1
    model = regression_model()

    # Fit the model on the train set
    print('\n\n\nTraining Model # ' , i+1 , '\n\n') # Print the Model Number that is being trained
    model.fit(X_train, Y_train, validation_split=0.3, epochs=100)
    print('\n')
    
    # Make prediction on the test set
    Y_predicted = model.predict(X_test)
    
    # Calculate the mean square error
    mse = mean_squared_error(Y_test, Y_predicted)
    
    # Add the mse to the list_of_mse list
    list_of_mse.append(mse)

end_time = datetime.now() # Ending time of the for loop execution

# Print time taken for fitting 50 models and calucating the Mean and Standard Deviation of MSE of 50 models
print('\n\nTotal Execution Time : ' , format(end_time - start_time))
    

Streaming output truncated to the last 5000 lines.
16/16 [==============================] - 0s 6ms/step - loss: 141.2429 - val_loss: 160.9287
Epoch 100/100
16/16 [==============================] - 0s 5ms/step - loss: 125.0198 - val_loss: 160.5157





Training Model #  27 


Epoch 1/100
16/16 [==============================] - 1s 19ms/step - loss: 19407.2869 - val_loss: 10085.4082
Epoch 2/100
16/16 [==============================] - 0s 5ms/step - loss: 8477.6473 - val_loss: 4508.5571
Epoch 3/100
16/16 [==============================] - 0s 5ms/step - loss: 3884.2155 - val_loss: 2105.8108
Epoch 4/100
16/16 [==============================] - 0s 5ms/step - loss: 1757.9859 - val_loss: 1032.5752
Epoch 5/100
16/16 [==============================] - 0s 5ms/step - loss: 854.7585 - val_loss: 567.6989
Epoch 6/100
16/16 [==============================] - 0s 5ms/step - loss: 576.0744 - val_loss: 395.2064
Epoch 7/100
16/16 [==============================] - 0s 5ms/step - loss: 408.8812 - val_loss: 3

In [38]:
# Calculate the Mean of the MSE of 50 models
mean_of_mse = stats.mean(list_of_mse)

# Calculate the Standard Deviation of the MSE of 50 models
std_of_mse = stats.stdev(list_of_mse)

# Print the Mean and Standard Deviation of MSE of 50 models
print('\n\nMean of the MSE of 50 Models : ' , str(mean_of_mse))
print('Standard Deviation of MSE of 50 Models : ' , str(std_of_mse))



Mean of the MSE of 50 Models :  132.55491108188656
Standard Deviation of MSE of 50 Models :  84.06283205976719


<br/>

#### <font color = green> Comparision of Mean of MSE with Mean of MSE from PART B </font>
<table style="width:20%">
  <tr>
    <th>Mean of MSE of PART A</th>
    <th>Mean of MSE of PART B</th>
    <th>Mean of MSE of PART C</th>
  </tr>
  <tr>
    <td>212.13</td>
    <td>184.57</td>
    <td>132.55</td>
  </tr>
</table>

The table above compares the Mean of **MSE for PART A**, **Mean of MSE for PART B** and **Mean of MSE for PART C**. As can be seen, the value of Mean of MSE of PART C is marginally larger than that of PART B. This shows that the effect of **normalizing the features** as well as **increasing the number of epochs by 2** did not improve the performance of the regression model and helps it in finding the line of best fit

<br/>

# <font color = blue> END OF PART C </font>

<hr/>

<br/>

<br/>

<br/>

<hr/>

# <font color = blue> PART D : BASELINE MODEL WITH INCREASED HIDDEN LAYERS </font>


In this part, all the tasks from <b>PART B</b> are performed, but this time the number of hidden layers are increased to 3

<b>The new model will have : </b>
<ul>
        <li> Input layer with 10 nodes </li>
        <li> 3 hidden layers, each with 10 nodes and ReLU activation function </li>
        <li> Adam optimizer and mean squared error loss function </li>
</ul>


<p/>

<p/>

## <font color = darkorange>Task 1 : Train and Test the Baseline Model with 100 epochs and Increased Hidden Layers</font>

In order to train and test the the baseline model with normalized features, 100 epochs and increased hidden layers, the following steps are performed :
<ol>
    <li>Normalize the features (X)</li>
    <li>Randomly split the data into <i>X_train, X_test, Y_train, Y_test</i> sets</li>
    <li>Create a new model with 100 epochs</li>
    <li>Train the model using <i>X_train, Y_train</i> using <b>50</b> epochs</li>
    <li>Evaluate the model using <i>X_test, Y_test</i></li>
    <li>Get the predictions on the <i>X_test</i> set </li>
    <li>Compute the <b>mean squared error</b> on the test set using sklearn</li>
</ol>

### <font color = #2980B9> Step 1 : Normalize the features (X) </font>

<b>Note</b> : As the features (X) have already been normalized the features (X), hence this part is skipped

### <font color = #2980B9> Step 2 : Randomly split the data into <i>X_train, X_test, Y_train, Y_test </font>

In [39]:
# Creating X_train, X_test, Y_train and Y_test sets
X_train, X_test, Y_train, Y_test = data_split()

### <font color = #2980B9> Step 3 : Create a new regression model with 3 hidden layers, each with 10 nodes and ReLU activation  </font>

In [40]:
def three_layer_regression_model () :
    
    # Create the model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(X.shape[1],)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))

    # Compile the model
    model.compile(optimizer='adam', loss='mean_squared_error')
    
    return model
    

### <font color = #2980B9> Step 4 : Train the model using <i>X_train, Y_train</i> using 50 epochs </font>

<b>Note </b> : Since the regression model has already been created using `def regression_model` in <b>TASK 1</b> and there are no changes being made to it, hence there is no need to write the code for the model again. 

In [41]:
model = regression_model()

# Fit the model on the train set
model.fit(X_train, Y_train, validation_split=0.3, epochs=100)

Epoch 1/100
16/16 [==============================] - 1s 13ms/step - loss: 13157.8243 - val_loss: 2299.0500
Epoch 2/100
16/16 [==============================] - 0s 4ms/step - loss: 1881.2792 - val_loss: 1679.4698
Epoch 3/100
16/16 [==============================] - 0s 4ms/step - loss: 1680.5527 - val_loss: 1218.0470
Epoch 4/100
16/16 [==============================] - 0s 4ms/step - loss: 1194.1488 - val_loss: 1033.3153
Epoch 5/100
16/16 [==============================] - 0s 4ms/step - loss: 967.5937 - val_loss: 884.7729
Epoch 6/100
16/16 [==============================] - 0s 4ms/step - loss: 900.6545 - val_loss: 811.4247
Epoch 7/100
16/16 [==============================] - 0s 4ms/step - loss: 888.2007 - val_loss: 755.4598
Epoch 8/100
16/16 [==============================] - 0s 4ms/step - loss: 860.8392 - val_loss: 700.1383
Epoch 9/100
16/16 [==============================] - 0s 4ms/step - loss: 724.7038 - val_loss: 649.0769
Epoch 10/100
16/16 [==============================] - 0s 4ms/st

### <font color = #2980B9> Step 5 : Get the predictions on the X_test </font>

In [42]:
# Store the predictions in a variable Y_Predicted
Y_predicted = predict()

### <font color = #2980B9> Step 6 : Compute the <i>mean squared error</i> on the test set using sklearn </font>

In [43]:
# Calculate the mean square error

mse = calculate_mse()
print('Mean Square Error (MSE) of the Baseline Model with Normalized Features is : ' , str(mse))

Mean Square Error (MSE) of the Baseline Model with Normalized Features is :  103.483370755894


<br/>

<p/>

## <font color = darkorange>Task 2 : Create 50 Models and Calculate the <i>µ</i> & <i>σ</i> of their MSE Errors with new Features</font>

In order to train 50 models with new features (normalized features) and calulate the mean (µ) and standard deviation (σ) of their mean square errors (MSE) the following steps are performed :
<ol>
    <li>Create an empty list <code>list_of_mse</code> to store the mean square error of each of the models</li>
    <li>Define a for loop and perform each of the following steps in the loop</li>
        <ol>
        <li>Randomly split the data into <i>X_train, X_test, Y_train, Y_test</i> sets</li>
        <li>Train the model using <i>X_train, Y_train</i> using <b>50</b> epochs</li>
        <li>Evaluate the model using <i>X_test, Y_test</i></li>
        <li>Get the predictions on the <i>X_test</i> set </li>
        <li>Compute the <b>mean squared error</b> on the test set using sklearn</li>
    </ol>
</ol>

### <font color = #2980B9> Step 1 : Creating the <i>list_of_mse</i> list</font>

In [44]:
# Create the empty lists
list_of_mse = []

### <font color = #2980B9> Step 2 : Creating 50 Models, Calculating The MSE for Each and <i>µ</i> & <i>σ</i> of the 50 MSE Values in <i>list_of_mse</i> </font>

In [45]:
# Create the for loop to split the data, create, compile & fit model, evaluate & nake predictions, caluclate mse and store
# in list_of_mse

start_time = datetime.now() # Starting time of the for loop execution

for i in range(50) :
    # Split the data into train and test set
    X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3)
    
    # Create and compile the regression model using the function regression_model as defined in TASK 1
    model = regression_model()

    # Fit the model on the train set
    print('\n\n\nTraining Model # ' , i+1 , '\n\n') # Print the Model Number that is being trained
    model.fit(X_train, Y_train, validation_split=0.3, epochs=100)
    print('\n')
    
    # Make prediction on the test set
    Y_predicted = model.predict(X_test)
    
    # Calculate the mean square error
    mse = mean_squared_error(Y_test, Y_predicted)
    
    # Add the mse to the list_of_mse list
    list_of_mse.append(mse)

end_time = datetime.now() # Ending time of the for loop execution

# Print time taken for fitting 50 models and calucating the Mean and Standard Deviation of MSE of 50 models
print('\n\nTotal Execution Time : ' , format(end_time - start_time))
    

Streaming output truncated to the last 5000 lines.
16/16 [==============================] - 0s 7ms/step - loss: 113.2738 - val_loss: 115.8358
Epoch 100/100
16/16 [==============================] - 0s 6ms/step - loss: 109.0145 - val_loss: 117.7740





Training Model #  27 


Epoch 1/100
16/16 [==============================] - 1s 16ms/step - loss: 1586.9627 - val_loss: 759.0740
Epoch 2/100
16/16 [==============================] - 0s 7ms/step - loss: 793.9290 - val_loss: 459.2501
Epoch 3/100
16/16 [==============================] - 0s 6ms/step - loss: 485.5961 - val_loss: 377.8925
Epoch 4/100
16/16 [==============================] - 0s 6ms/step - loss: 394.8160 - val_loss: 312.0142
Epoch 5/100
16/16 [==============================] - 0s 6ms/step - loss: 341.6868 - val_loss: 280.9560
Epoch 6/100
16/16 [==============================] - 0s 6ms/step - loss: 304.9843 - val_loss: 259.7516
Epoch 7/100
16/16 [==============================] - 0s 5ms/step - loss: 283.2035 - val_loss: 243.6577
E

In [46]:
# Calculate the Mean of the MSE of 50 models
mean_of_mse = stats.mean(list_of_mse)

# Calculate the Standard Deviation of the MSE of 50 models
std_of_mse = stats.stdev(list_of_mse)

# Print the Mean and Standard Deviation of MSE of 50 models
print('\n\nMean of the MSE of 50 Models : ' , str(mean_of_mse))
print('Standard Deviation of MSE of 50 Models : ' , str(std_of_mse))



Mean of the MSE of 50 Models :  117.63979555900275
Standard Deviation of MSE of 50 Models :  36.43211651176892


<br/>

#### <font color = green> Comparision of Mean of MSE with Mean of MSE with PART C </font>
<table style="width:30%">
  <tr>
    <th>Mean of MSE of PART A</th>
    <th>Mean of MSE of PART B</th>
    <th>Mean of MSE of PART C</th>
    <th>Mean of MSE of PART D</th>
  </tr>
  <tr>
    <td>212.13</td>
    <td>184.57</td>
    <td>132.55</td>
    <td>117.63</td>
  </tr>
</table>

The table above compares the Mean of **MSE for PART A**, **Mean of MSE for PART B**, **Mean of MSE for PART C** and **Mean of MSE for PART D**. As can be seen, the value of Mean of MSE of PART D is marginally smaller than that of PART C and is the smallest value obtained. This shows that the effect of **normalizing the features** as well as **increasing the number of epochs by 2** yield the best results in terms of the performance of the regression model and helps it in finding the line of best fit

<br/>

# <font color = blue> END OF PART D</font>

<hr/>

<br/>

# <font color = ac36e3> END NOTE </font>

Although the results above table show that the best performance is achieved by normalizing the features, increasing the number of epochs **and** increasing the number of hidden layers, this might not be decisive. Repeating ***TASK 2*** for **PART A**, **PART B**, **PART C** and **PART D** several times shows different results. However, for the purposes of this project, those results are not included. 

<br/>

<hr/>

# <font color = red> END OF NOTEBOOK </font>